Improvement backports from CDT_3 branch (Follow-up to PR #8170)

lrineau opened the pull request:
## Summary of Changes

In the PR #8170, merged for CGAL-6.0, there was several commits that I had to revert, because they broke the testsuite results:

- [x] commit 612d6e3f592a133a21d14a00e0ffe849b11e2b04, that modified the function `save_binary_file`. That was no correct as regards to the C3t3 concept. It is now fixed by commit https://github.com/CGAL/cgal/pull/8273/commits/c10dcf7daf6ba7fa8b382043a220430618492a9b from this PR.
- [x] commit 1ca6c17a2a165dccb2204f01d38d6d3d903d11c9, that optimized the efficiency of `Polyline_constraint_hierarchy_2` using an unordered map, instead of a `std::map`. The issue was that is broke the determinism of Mesh_2. See https://github.com/CGAL/cgal/pull/8273#discussion_r1638397885.
- [ ] commit ac47f30cd8a0921e52e3bd2a88fb435092eef837, that introduced an experimental tuple protocol for `CGAL::Segment_3`, for some CGAL kernels.
- [x] commit df28ddc43ab3af1a54a2d9ad9587ab846707b752, that fixed a CMake warning about VTK-9.x. Now fixed by another PR (#8279).

This pull-request reintroduces those features, for CGAL-6.1.

## Release Management

* Affected package(s):
* License and copyright ownership: maintenance by GeometryFactory

lrineau committed:
use boost::unordered_flat_map to optimize Polyline_constraint_hierarchy_2

lrineau committed:
allow to call save_binary_file on non-C3t3 type (like a CDT_3)

lrineau committed:
Mesh_2: sort the sequence tr.subconstraints_begin(), tr.subconstraints_end()

lrineau commented:
## Order of the sequence of values from the range `CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()`

@janetournois @afabri

- The patch https://github.com/CGAL/cgal/pull/8273/commits/93fd96644c9e2a1aab33433880afada3e4eed02c makes the range [`CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()`](https://doc.cgal.org/6.0/Triangulation_2/classCGAL_1_1Constrained__triangulation__plus__2.html#af25114a7e1675194367f8f9de9de90d2) non-deterministic: instead of a `std::map` with a custom `compare` functor, now the sub-constraints are in an *unordered* map, for efficiency reasons. Note that the range was never documented as sorted or deterministic.
- As my patch had broken a test from Mesh_2 (the test `test_meshing`), I had to fix that non-determinism. That was complicated to fix directly in `Constrained_triangulation_plus_2` or `Constrained_triangulation_plus_2`. Instead, I have chosen to fix the only place in CGAL where the order of that range mattered, in `Mesh_2/include/CGAL/Mesh_2/Refine_edges.h`. See the commit https://github.com/CGAL/cgal/pull/8273/commits/82b53596fd6beaaf6c5accf2182cbdd3553b3365.

Is that correct? That is a sort of breaking change, but the ordering of `Ctp_2::constraints()` was never documented.

lrineau commented:
> ## Order of the sequence of values from the range `CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()`

Jane (IRL) suggested that:

  -  it is documented that there is a breaking change
  -  and add a Tag to `Constrained_triangulation_plus_2` to allow users to fallback to the previous behavior.

lrineau commented on Triangulation_2/include/CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h:
```diff
@@ -171,8 +162,11 @@ class Polyline_constraint_hierarchy_2
   typedef std::set<Constraint_id>           Constraint_set;
-  typedef std::map<Edge, Context_list*,
-                   Pair_compare>            Sc_to_c_map;
+#if CGAL_USE_BOOST_UNORDERED
+  typedef boost::unordered_flat_map<Edge, Context_list*, boost::hash<Edge>> Sc_to_c_map;
+#else
+  typedef std::unordered_map<Edge, Context_list*, boost::hash<Edge>> Sc_to_c_map;
+#endif
```
## Order of the sequence of values from the range `CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()`

@janetournois @afabri

- The patch https://github.com/CGAL/cgal/pull/8273/commits/93fd96644c9e2a1aab33433880afada3e4eed02c makes the range [`CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()`](https://doc.cgal.org/6.0/Triangulation_2/classCGAL_1_1Constrained__triangulation__plus__2.html#af25114a7e1675194367f8f9de9de90d2) non-deterministic: instead of a `std::map` with a custom `compare` functor, now the sub-constraints are in an *unordered* map, for efficiency reasons. Note that the range was never documented as sorted or deterministic.
- As my patch had broken a test from Mesh_2 (the test `test_meshing`), I had to fix that non-determinism. That was complicated to fix directly in `Constrained_triangulation_plus_2` or `Constrained_triangulation_plus_2`. Instead, I have chosen to fix the only place in CGAL where the order of that range mattered, in `Mesh_2/include/CGAL/Mesh_2/Refine_edges.h`. See the commit https://github.com/CGAL/cgal/pull/8273/commits/82b53596fd6beaaf6c5accf2182cbdd3553b3365.

Is that correct? That is a sort of breaking change, but the ordering of `Ctp_2::constraints()` was never documented.

lrineau replied:
Jane (IRL) suggested that:

  -  it is documented that there is a breaking change
  -  and add a Tag to `Constrained_triangulation_plus_2` to allow users to fallback to the previous behavior.

lrineau replied:
(I have copy-pasted my previous *comment* into a *discussion*. So that all the messages about this discussion are at the same place.)

sloriot replied:
In order to avoid ping-pong changes, I suggest to have `CGAL::unordered_map` and to use it everywhere (at least for default hashable items).

lrineau replied:
## Speedups

@afabri:
> Any idea how big the gain in efficiency is @lrineau ?

One the data set ["Thingi #101881"][https://ten-thousand-models.appspot.com/detail.html?file_id=101881], I have run the experimental CDT_3 code, with compilation flags `-O3 -DCGAL_NDEBUG`. The timings of the step "conforming of facets borders" vary a lot:

| with...                     | time    | factor | 
|----------------------------:|---------|-------:|
| `std::map`                  | 4110 ms |  1.00  |
| `std::unordered_map`        | 3623 ms |  0.88  |
| `boost::unordered_flat_map` | 2897 ms |  0.70  |

I have repeated multiple time each of the three runs. The variance of the timings between each run is less than 5%, and the difference between the three implementations is about 15%.

And the code the "CDT_3 conforming" does a lot more than just handling the `CGAL::Constrained_triangulation_plus_2` data-structure. The speedups are significant, in my opinion.

## Code

@sloriot:
> I suggest to have `CGAL::unordered_map`

Indeed. I have a patch. I will push it...

afabri replied:
We had several times users where we worked hard to make algorithms deterministic for reproductibility. So this is maybe more important than speed.  Also I find it more natural that subconstraints are ordered along the constraint.

afabri replied:
I just realize that subconstraints are not the subconstraints of a constraint, but all subconstraints in the constrained triangulation.    Instead iterating over the Xmap, why don't we iterate over the vertices of the constraints, as a subconstraint is just a pair of consecutive vertices on a constraint?

lrineau replied:
## `CGAL::unordered_map`

I have pushed the commit 552685784b6705ea0a3b8f4b0a806de0a8042cc7. The results of the testsuites of `Triangulation_2`, `Mesh_2`, and `Polyline_simplification_2` are PASSED on my machine (with Boost.Unordered available).

lrineau replied:
`CGAL::Constrained_triangulation_plus_2<Tr>::subconstraints()` is still non-deterministic. And it was never documented as such.

The new range `CGAL::Constrained_triangulation_plus_2<Tr>::edges()` is deterministic.

lrineau commented:
Test a comment that will not trigger `build_doc.yml`...

lrineau commented:
/build:test_build_doc.yml

lrineau commented:
Another test comment that will not trigger build_doc.yml...

lrineau committed:
Merge remote-tracking branch 'cgal/master' into pr/lrineau/8273

lrineau committed:
Add CGAL::unordered_flat_map implementation

...and `refactor Polyline_constraint_hierarchy_2` to use it.

`CGAL::unordered_flat_map` will be Boost `unordered_flat_map` if availlable, or the standard `std::unordered_map` otherwise.

lrineau committed:
fix an error detected by UBSAN

lrineau committed:
fix CGAL_USE_BARE_STD_MAP with Polyline_constraint_hierarchy_2

lrineau committed:
new allocator type, that allocates randomly on purpose

To debug non-determinism on Linux platforms.

lrineau committed:
fix a missing #include

sloriot committed:
TWS

lrineau committed:
cleanup existing code

lrineau committed:
Boost >= 1.74 is now required

lrineau committed:
CDT_plus_2 hierarchy: add Edge_iterator, that is deterministic

That uses Boost.STLInterfaces from Boost >= 1.74.

lrineau commented on Triangulation_2/include/CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h:
```diff
@@ -262,18 +413,42 @@ class Polyline_constraint_hierarchy_2
     return sc_to_c_map.end();
   }
 
+  Edge_iterator edges_begin() const {
+    BOOST_STL_INTERFACES_STATIC_ASSERT_CONCEPT(Edge_iterator, std::bidirectional_iterator);
+    BOOST_STL_INTERFACES_STATIC_ASSERT_ITERATOR_TRAITS(Edge_iterator, std::bidirectional_iterator_tag,
+                                                       std::bidirectional_iterator, Edge, Edge,
+                                                       typename Edge_iterator::pointer, std::ptrdiff_t);
+    return Edge_iterator(Edge_iterator::Construction_access::begin_tag(), &constraint_set);
+  }
+
+  Edge_iterator edges_end() const {
+    return Edge_iterator(Edge_iterator::Construction_access::end_tag(), &constraint_set);
+  }
+
+  auto edges() const { return Iterator_range<Edge_iterator>(edges_begin(), edges_end()); }
+
```
@afabri the range `edges()` is a range on `std::pair<Vertex_handle, Vertex_handle>`, one for earch subconstraints.

Suggestion for a better name?

afabri replied:
As `Edge` is a pair of face handle and an index it should not be `Edge_iterator`but  `Subconstraint_iterator`.

lrineau committed:
hierarchy: remove  `remove_Steiner(T v, T va, T vb)`

  That code was never used nor tested, and cannot compile anyway.

lrineau committed:
hierarchy: remove two undefined member functions

lrineau committed:
major cleanup of Polyline_constraint_hierarchy_2

- remove all mentions of `Edge` and `Constraint`
- `Subconstraint_iterator` is renamed `Subconstraint_and_contexts_iterator` (because of its value type)
- a new `Subconstraint_iterator`, with value type `Subconstraint`
- a few unused/untested and uncompilable functions are removed from the code
- a lot of internal renamings

== Breaking changes ==

For `Constrained_triangulation_plus_2`, there are a few breaking changes...

- The value type of `subconstraints_begin()`, `subconstraints_end()`, of the range `subconstraints()` has changed to `Subconstraint` (a simple `std::pair` of vertex handles). That is actually a kind of bug-fix, because it was documented as such in the user manual.
- The new member functions `subconstraints_and_contexts_begin()`, `subconstraints_and_contexts_end()`, `subconstraints_and_contexts()` are created get the old value type (`std::pair<const Subconstraint, std::list<Context>*>`).
- A few range types have changed from `CGAL::Iterator_range<It>` to `unspecified_type`, for efficiency reasons.
- Doc fixes.

== Determinism ==

Even if it was not documented, the range `subconstraints()` is deterministic (used by Mesh_2), and `subconstraints_and_contexts()` is not.

sloriot commented:
boost 1.74 is requested but some test plateforms do not have it. Do you really need that minimal version?

lrineau committed:
doc bug-fix

lrineau committed:
fix protection against min/max macros

The regular expression I used was:

```
((?!(?:^.*(\/\/|\/\*).*|^ *\* .*|^[^"]*"(?:"[^"]*"|[^"])*))^(?:.*[ ,\(]|))(\b(?:(?:[A-Za-z]+::)*)(?:max|min))\b *\(
```

lrineau commented:
## Dump the requirement for Boost to 1.74 or later? (currently 1.72)

> boost 1.74 is requested but some test plateforms do not have it. Do you really need that minimal version?

### Short answer

Yes, I really need it.


### Long answer

Actually no: I could use the ancient [`boost::iterator_facade` from Boost.Iterator](https://www.boost.org/doc/libs/1_87_0/libs/iterator/doc/iterator_facade.html), but [Boost.STLInterfaces](https://www.boost.org/doc/libs/release/libs/stl_interfaces) is a lot nicer to use, and create release iterator types, even with proxy iterators like the one I wrote.

But platforms supported by CGAL-6.0 currently have also Boost 1.74 or later.

<details><summary>Details... (click to expand)</summary>

  - Fedora has Boost>=1.74 [since Fedora 35](https://src.fedoraproject.org/rpms/boost/c/dd07b4b21f480f41a031ff3a352ef4e556959433), published 2021-11-02, and the [previous version Fedora 34 is not supported since 2021-11-02](https://en.wikipedia.org/wiki/Fedora_Linux_release_history) (the current version is Fedora 41, two versions per years).
  - Debian has Boost>=1.74 [since its version `bullseyes`](https://packages.debian.org/search?searchon=names&keywords=libboost-dev), that is currently the "oldstable" version of Debian. And the [version before, `buster`, is](https://www.debian.org/releases/index.html):
    
    - "EOL" since 2022-09-10,
    - "EOL LTS" since 2024-06-30.
    
  - Ubuntu: [since Ubuntu 22.04 LTS (aka _Focal_)](https://launchpad.net/ubuntu/+source/boost-defaults), and the previous supported version, 20.04 LTS, does not have Boost 1.72 anyway,
  - Same for Redhat Enterprise/CentOS/AlmaLinux: [version 9 has 1.75, and the previous version had only Boost 1.66](https://repology.org/project/boost/versions#centos_8).
  - Conan [has Boost 1.86](https://conan.io/center/recipes/boost), Conda [has Boost 1.85](https://anaconda.org/conda-forge/boost), vcpkg [has Boost 1.86](https://vcpkg.link/ports/boost), Homebrew has [Boost 1.87](https://formulae.brew.sh/formula/boost)... all package managers have very recent versions.

So... 

  - there is no supported versions of Linux that is currently supported by CGAL-6.0 and will not be by CGAL-6.1,
  - all the package managers have Boost >= 1.85,
  - and for Windows that is as easy to install any version of Boost as it is for version 1.72.

</details>

lrineau commented on Installation/CHANGES.md:
```diff
@@ -3,6 +3,9 @@
+### General Changes
+- The minimal supported version of Boost is now 1.74.0.
+
 ### [Algebraic Kernel](https://doc.cgal.org/6.1/Manual/packages.html#PkgAlgebraicKernelD)
```
Must also announce the breaking changes in `Triangulation_2/doc/Triangulation_2/CGAL/Constrained_triangulation_plus_2.h`.

lrineau commented:
/build:doc

lrineau committed:
fix compilation errors with C++<20

lrineau committed:
document the breakng change

lrineau committed:
doc fix: add that iterators are all bidirectional

lrineau committed:
rewrite/refactor a lot of the code

lrineau committed:
add failing tests (for now)

lrineau committed:
fix the failing test

lrineau committed:
cleanup in CGAL/Iterator_range.h

... and add a default constructor to create an
empty range (with singular iterators).

lrineau committed:
Polyline_constraint_hierarchy_2: get_contexts -> context

lrineau committed:
clearnup: remove undocumnted/unused member functions

lrineau committed:
Polyline_constraint_hierarchy_2: more refactoring

- refactor the member function `enclosing_constraint`
- refactor things around `sc_to_sc_map`, using new member functions
  to encapsulate the logic
- remove unused/undocuments functions

lrineau committed:
rename contexts_end to contexts_not_found because that name is already used in the class

lrineau committed:
add a nested type using Vertex_handle_compare = Compare;

... add switch to `using` in the whole file.

lrineau committed:
use using-declarations and type aliases

lrineau committed:
add CGAL::unordered_flat_set

lrineau committed:
virtual -> override

lrineau committed:
derive Ct_plus_2 from the hierarchy

Instead of storing the hierarchy as a data member, store it
as a hidden (protected) base class. That allows to forward
member functions easily with using-declarations. That also avoids
mismatches between the names in the triangulation and the hierarchy.

lrineau committed:
whitespace and comments

lrineau committed:
glue constraints_set.insert(..) and new_constraint_id()

.. to prepare a refactoring

lrineau committed:
CDt_plus_2: lots of refactoring

- encapsulate the creation/erasure of constraints in the hierarchy class
- add encapsulation in the hierarchy class, to detect direct uses of members
- add a test of `CDt_plus_2::insert_vertex_in_constraint`, and fix it
- rename `concatenate` and `concatenate2` to
  - `concatenate`
  - `prepend`
  and force a rvalue reference on the argument corresponding to the constraint
  that will be swallowed
- rename `split` to `split_tail` and `split2` to `split_head`
- rename `fix_contexts` to give it a longer self-explaning name

lrineau committed:
fix a warning from UBSAN (in a postcondition)

lrineau committed:
recycle the constraints indices

sloriot commented on Triangulation_2/include/CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h:
```diff
@@ -1055,189 +1192,162 @@ split_constraint(T va, T vb, T vc){
+
+  auto disp_vertex = [&vertex_num_map](Vertex_handle v) {
+    return CGAL::IO::oformat(
+        [v, &vertex_num_map](auto& os) -> decltype(os)& {
```
```suggestion
        [v](auto& os) -> decltype(os)& {
```

lrineau replied:
I must reject this suggestion, because `vertex_num_map` is used, line 1310, in case the vertex type does not have timestamps. I will add a `[[maybe_unused]]`.

lrineau committed:
small optimization

lrineau committed:
fix compilation errors with MSVC 2017 (VC++ 19.16)

lrineau committed:
fix another compilation error

VC++ 19.16 thought there was an ambiguous call to `operator==`
between `skip_iterator` and `skip_iterator` in the internals of `boost::iterator_adaptor` (yet another matching bug) of VC++ 19.16).

I solved it by using the more modern `boost::stl_interfaces::iterator_interface` from Boost.STLintface (Boost>=1.74).

afabri commented:
more errors:  
````

    35>C:\CGAL_ROOT\CGAL-6.1-Ic-73\include\CGAL/Constrained_Delaunay_triangulation_2.h(912): error C2244: 'CGAL::Constrained_Delaunay_triangulation_2<Gt,Tds_,Itag_>::virtual_insert': unable to match function definition to an existing declaration [C:\CGAL_ROOT\CGAL-6.1-Ic-73\cmake\platforms\MSVC2017-Release-64bits\test\Triangulation_2\issue_3447.vcxproj]
         C:\CGAL_ROOT\CGAL-6.1-Ic-73\include\CGAL/Constrained_Delaunay_triangulation_2.h(907): note: see declaration of 'CGAL::Constrained_Delaunay_triangulation_2<Gt,Tds_,Itag_>::virtual_insert'
         C:\CGAL_ROOT\CGAL-6.1-Ic-73\include\CGAL/Constrained_Delaunay_triangulation_2.h(912): note: definition
         C:\CGAL_ROOT\CGAL-6.1-Ic-73\include\CGAL/Constrained_Delaunay_triangulation_2.h(912): note: 'Constrained_Delaunay_triangulation_2<Gt,Tds_,Itag_>::Vertex_handle CGAL::Constrained_Delaunay_triangulation_2<Gt,Tds_,Itag_>::virtual_insert(const Constrained_Delaunay_triangulation_2<Gt,Tds_,Itag_>::Geom_traits::Point_2 
````

lrineau committed:
extra fix for MSVC 2017 in Constrained_Delaunay_triangulation_2

lrineau committed:
fix compilation error with gcc-12.2

lrineau committed:
fix the detection of <format> and std::format

afabri commented:
```
Meshing the triangulation with size 0...C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.16.27023\include\xtree(275) : Assertion failed: map/set iterators incompatible
```
[here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-74/Mesh_2/TestReport_Christo_MSVC2017-Debug-64bits.gz)

afabri commented:
```
/mnt/testsuite/include/CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h:129:61: error: static assertion failed
  129 |   BOOST_STL_INTERFACES_STATIC_ASSERT_CONCEPT(Point_it, std::bidirectional_iterator);
```
[here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-74/Mesh_2/TestReport_lrineau_Ubuntu-GCC_master_CXX20-Release.gz)

afabri commented:
```
include\CGAL/Base_with_time_stamp.h(23): error C2248: 'CGAL::Constrained_triangulation_face_base_2<K,CGAL::Triangulation_face_base_2<Gt,CGAL::Triangulation_ds_face_base_2<void>>>::Base': cannot access private typedef declared in class 'CGAL::Constrained_triangulation_face_base_2<K,CGAL::Triangulation_face_base_2<Gt,CGAL::Triangulation_ds_face_base_2<void>>>'
```
[here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-74/Triangulation_2/TestReport_Christo_MSVC2017-Debug-64bits.gz)

lrineau committed:
fix compilation errors with C++>=20, or g++12.2.0

lrineau committed:
fix for STL debug mode

It is invalid to try to detect if an iterator is singular or value-initialized.
The only operations allowed with value-initialized iterators are:
  - copy the iterator, or
  - destroy or assign the iterator.

Comparisons like `vertex_it == Vertex_it{}` are not allowed.

lrineau commented:
> ```
> Meshing the triangulation with size 0...C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.16.27023\include\xtree(275) : Assertion failed: map/set iterators incompatible
> ```
> 
> [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-74/Mesh_2/TestReport_Christo_MSVC2017-Debug-64bits.gz)

The code was doing bad operations with value-initialized iterators. Fixed in d837dbde218d9c24727b65b8e533404651341954.


> ```
> /mnt/testsuite/include/CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h:129:61: error: static assertion failed
>   129 |   BOOST_STL_INTERFACES_STATIC_ASSERT_CONCEPT(Point_it, std::bidirectional_iterator);
> ```
> 
> [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-74/Mesh_2/TestReport_lrineau_Ubuntu-GCC_master_CXX20-Release.gz)

With C++20, the behavior of Boost.STLinterface was different. I have fixed, and tested with and without C++20. And I found a bug in the macro `BOOST_STL_INTERFACES_STATIC_ASSERT_ITERATOR_TRAITS` before Boost 1.83.0. That is fixed in 571c2ccadc440672917313d8540016894ae7fa78.



> ```
> include\CGAL/Base_with_time_stamp.h(23): error C2248: 'CGAL::Constrained_triangulation_face_base_2<K,CGAL::Triangulation_face_base_2<Gt,CGAL::Triangulation_ds_face_base_2<void>>>::Base': cannot access private typedef declared in class 'CGAL::Constrained_triangulation_face_base_2<K,CGAL::Triangulation_face_base_2<Gt,CGAL::Triangulation_ds_face_base_2<void>>>'
> ```
> 
> [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-74/Triangulation_2/TestReport_Christo_MSVC2017-Debug-64bits.gz)

A known issue with MSVC 2017 (see [the repro](https://compiler-explorer.com/z/d5hxv7Wze)). Fixed in 571c2ccadc440672917313d8540016894ae7fa78 as well (it should have been a distinct commit).

lrineau committed:
another fix for the buggy compiler from MSVC 2017

sloriot commented:
- Timeout [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-76/Mesh_3_Examples/TestReport_lrineau_Fedora-rawhide-Release.gz)
- Timeout [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-76/Poisson_surface_reconstruction_3_Examples/TestReport_lrineau_Fedora.gz)
-  ICE [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-76/Triangulation_2_Examples/TestReport_Christo_MSVC2017-Release-64bits.gz) and [there](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.1-Ic-76/Triangulation_2_Examples/TestReport_Christo_MSVC2017-Debug-64bits.gz)

Note that it might come from the time_stamp PR.

lrineau committed:
add CMake option CGAL_with_benchmarks

lrineau committed:
fix WKT I/O error handling

lrineau committed:
add a benchmark for Polyline_simplification_2

Using the tool `compare.py` from [Google Benchmark Tools][1]...

[1]: https://github.com/google/benchmark/blob/main/docs/tools.md

...given that:

  - `benchmark_simplify-AFTER` is the file `benchmark_simplify.cpp`
    compiled from this branch, and flags `-msse3  -DNDEBUG -O3`
    with gcc version 14.2.1,
  - `benchmark_simplify-BEFORE` is the same file
    compiled from the branch `master`, and same flags and compiler.

the command line was:

```shell
compare.py benchmarks build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-BEFORE build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-AFTER Data/data/wkt/norway-MP.wkt --benchmark_repetitions=20 --benchmark_counters_tabular=true --benchmark_time_unit=s
```

The following results show that there is a speedup
of about 15% (mean) or 18% (median)

```
RUNNING: build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-BEFORE Data/data/wkt/norway-MP.wkt --benchmark_repetitions=20 --benchmark_counters_tabular=true --benchmark_time_unit=s --benchmark_out=/tmp/tmpvxyq_i24
2025-02-06T17:23:08+01:00
Running Data/data/wkt/norway-MP.wkt
Run on (16 X 800 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)
Load Average: 1.47, 2.07, 1.65
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations    #points  #polygons #polylines nb of constraints nb of sub-constraints nb of vertices
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
simplify file Data/data/wkt/norway-MP.wkt             0.067 s         0.067 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.066 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.065 s         0.065 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.073 s         0.072 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.066 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.067 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.071 s         0.071 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.074 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.066 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.075 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.074 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.076 s         0.076 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.078 s         0.078 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.074 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.067 s         0.067 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.069 s         0.069 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_mean        0.069 s         0.069 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_median      0.067 s         0.067 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_stddev      0.005 s         0.005 s            20          0          0          0                 0                     0              0
simplify file Data/data/wkt/norway-MP.wkt_cv           6.73 %          6.68 %            20      0.00%      0.00%      0.00%             0.00%                 0.00%          0.00%
RUNNING: build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-AFTER Data/data/wkt/norway-MP.wkt --benchmark_repetitions=20 --benchmark_counters_tabular=true --benchmark_time_unit=s --benchmark_out=/tmp/tmp59ui0dsw
2025-02-06T17:29:26+01:00
Running Data/data/wkt/norway-MP.wkt
Run on (16 X 800.961 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)
Load Average: 1.61, 1.94, 1.73
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations    #points  #polygons #polylines nb of constraints nb of sub-constraints nb of vertices
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.057 s         0.057 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.059 s         0.059 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.065 s         0.065 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.058 s         0.058 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.056 s         0.056 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.053 s         0.053 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.061 s         0.061 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.058 s         0.058 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.052 s         0.052 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.055 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.055 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.059 s         0.059 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.057 s         0.057 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.058 s         0.057 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.055 s         0.055 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.059 s         0.059 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_mean        0.057 s         0.057 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_median      0.057 s         0.056 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_stddev      0.003 s         0.003 s            20          0          0          0                 0                     0              0
simplify file Data/data/wkt/norway-MP.wkt_cv           5.37 %          5.32 %            20      0.00%      0.00%      0.00%             0.00%                 0.00%          0.00%
Comparing build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-BEFORE to build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-AFTER
Benchmark                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------
simplify file Data/data/wkt/norway-MP.wkt                       -0.1941         -0.1942             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1269         -0.1272             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.0900         -0.0918             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1049         -0.1065             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1204         -0.1206             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1279         -0.1282             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1962         -0.1961             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1370         -0.1377             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.0951         -0.0962             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1851         -0.1852             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2720         -0.2716             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1702         -0.1717             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2700         -0.2703             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2644         -0.2634             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2861         -0.2856             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2492         -0.2488             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2310         -0.2301             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1382         -0.1389             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1401         -0.1406             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1464         -0.1467             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
simplify file Data/data/wkt/norway-MP.wkt_mean                  -0.1804         -0.1806             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_median                -0.1547         -0.1549             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_stddev                -0.3458         -0.3479             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_cv                    -0.2018         -0.2041             0             0             0             0
OVERALL_GEOMEAN                                                 -0.1797         -0.1800             0             0             0             0
```

lrineau commented:
@sloriot

## Benchmarks

This PR given even bigger now. I had to:
  - fix the error handling of WKT input functions (765744e8b2ea77a00b2ba99b42f5b7e6527601c7),
  - change the CMake scripts to add `CGAL_with_benchmarks` as a CMake options (050677f002f7478d95a61d75c30db7c5cf27263d),
  - and then add a benchmark file using https://github.com/google/benchmark/ (bc44c91536c1ba39172d252f9a9855261a2b9f95)

### Results

Using `CGAL::Polyline_simplification_2::simplify` on a CDT constructed from the file `Data/data/wkt/norway-MP.wkt`, the execution from this PR is about 18% faster than the code from `master`.

The results can be seen at https://cgal.geometryfactory.com/~lrineau/simplify-bench.html
and in this screenshot:

![Screenshot_20250206_180048](https://github.com/user-attachments/assets/a27de40e-85a9-4505-9adf-41aa71307c26)


<details><summary>Details in text (click for more...)</summary>
<p>


Using the tool `compare.py` from [Google Benchmark Tools][1]...

[1]: https://github.com/google/benchmark/blob/main/docs/tools.md

...given that:

  - `benchmark_simplify-AFTER` is the file `benchmark_simplify.cpp`
    compiled from this branch, and flags `-msse3  -DNDEBUG -O3`
    with gcc version 14.2.1,
  - `benchmark_simplify-BEFORE` is the same file
    compiled from the branch `master`, and same flags and compiler.

the command line was:

```shell
compare.py benchmarks build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-BEFORE build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-AFTER Data/data/wkt/norway-MP.wkt --benchmark_repetitions=20 --benchmark_counters_tabular=true --benchmark_time_unit=s
```

The following results show that there is a speedup
of about 15% (mean) or 18% (median)

```
RUNNING: build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-BEFORE Data/data/wkt/norway-MP.wkt --benchmark_repetitions=20 --benchmark_counters_tabular=true --benchmark_time_unit=s --benchmark_out=/tmp/tmpvxyq_i24
2025-02-06T17:23:08+01:00
Running Data/data/wkt/norway-MP.wkt
Run on (16 X 800 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)
Load Average: 1.47, 2.07, 1.65
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations    #points  #polygons #polylines nb of constraints nb of sub-constraints nb of vertices
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
simplify file Data/data/wkt/norway-MP.wkt             0.067 s         0.067 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.066 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.065 s         0.065 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.073 s         0.072 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.066 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.067 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.071 s         0.071 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.074 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.066 s         0.066 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.075 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.074 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.076 s         0.076 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.078 s         0.078 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.074 s         0.074 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.067 s         0.067 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.064 s         0.064 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.069 s         0.069 s            12          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_mean        0.069 s         0.069 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_median      0.067 s         0.067 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_stddev      0.005 s         0.005 s            20          0          0          0                 0                     0              0
simplify file Data/data/wkt/norway-MP.wkt_cv           6.73 %          6.68 %            20      0.00%      0.00%      0.00%             0.00%                 0.00%          0.00%
RUNNING: build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-AFTER Data/data/wkt/norway-MP.wkt --benchmark_repetitions=20 --benchmark_counters_tabular=true --benchmark_time_unit=s --benchmark_out=/tmp/tmp59ui0dsw
2025-02-06T17:29:26+01:00
Running Data/data/wkt/norway-MP.wkt
Run on (16 X 800.961 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)
Load Average: 1.61, 1.94, 1.73
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations    #points  #polygons #polylines nb of constraints nb of sub-constraints nb of vertices
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.057 s         0.057 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.059 s         0.059 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.065 s         0.065 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.058 s         0.058 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.056 s         0.056 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.053 s         0.053 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.061 s         0.061 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.058 s         0.058 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.052 s         0.052 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.055 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.055 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.054 s         0.054 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.059 s         0.059 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.057 s         0.057 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.058 s         0.057 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.055 s         0.055 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt             0.059 s         0.059 s            13          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_mean        0.057 s         0.057 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_median      0.057 s         0.056 s            20          0        848          0               849               40.568k        40.565k
simplify file Data/data/wkt/norway-MP.wkt_stddev      0.003 s         0.003 s            20          0          0          0                 0                     0              0
simplify file Data/data/wkt/norway-MP.wkt_cv           5.37 %          5.32 %            20      0.00%      0.00%      0.00%             0.00%                 0.00%          0.00%
Comparing build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-BEFORE to build/Release/benchmark/Polyline_simplification_2/benchmark_simplify-AFTER
Benchmark                                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------
simplify file Data/data/wkt/norway-MP.wkt                       -0.1941         -0.1942             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1269         -0.1272             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.0900         -0.0918             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1049         -0.1065             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1204         -0.1206             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1279         -0.1282             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1962         -0.1961             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1370         -0.1377             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.0951         -0.0962             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1851         -0.1852             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2720         -0.2716             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1702         -0.1717             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2700         -0.2703             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2644         -0.2634             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2861         -0.2856             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2492         -0.2488             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.2310         -0.2301             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1382         -0.1389             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1401         -0.1406             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt                       -0.1464         -0.1467             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_pvalue                 0.0000          0.0000      U Test, Repetitions: 20 vs 20
simplify file Data/data/wkt/norway-MP.wkt_mean                  -0.1804         -0.1806             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_median                -0.1547         -0.1549             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_stddev                -0.3458         -0.3479             0             0             0             0
simplify file Data/data/wkt/norway-MP.wkt_cv                    -0.2018         -0.2041             0             0             0             0
OVERALL_GEOMEAN                                                 -0.1797         -0.1800             0             0             0             0
```

</p>
</details>

sloriot commented:
@lrineau could you please fix conflicts?

lrineau committed:
Merge branch 'master' into pr/lrineau/8273

lrineau committed:
fix a bug when `CGAL_BRANCH_BUILD` is falsy (in testsuite tarballs for example)

lrineau reviewed (commented):
Sorry for the mess in https://cgal.geometryfactory.com/CGAL/testsuite/results-6.1-Ic-81.shtml. It was because of this PR. Now fixed, and tested locally.

lrineau committed:
do not require `Boost_FOUND`
